src/adapter/breakpointPredictor.ts

```diff
@@ -52,126 +59,122 @@ export class BreakpointsPredictor {
-      return;
-    const fileUrl = urlUtils.absolutePathToFileUrl(absolutePath);
-    sourceMapUrl = urlUtils.completeUrl(fileUrl, sourceMapUrl);
-    if (!sourceMapUrl)
-      return;
-    if (!sourceMapUrl.startsWith('data:') && !sourceMapUrl.startsWith('file://'))
-      return;
-    try {
-      const map = await sourceUtils.loadSourceMap(sourceMapUrl, 0);
-      if (!map)
-        return;
-      this._sourceMapUrls.set(absolutePath, sourceMapUrl);
       for (const url of map.sources) {
         const sourceUrl = urlUtils.maybeAbsolutePathToFileUrl(this._predictor._rootPath, url);
-        const baseUrl = sourceMapUrl.startsWith('data:') ? fileUrl : sourceMapUrl;
         const resolvedUrl = urlUtils.completeUrlEscapingRoot(baseUrl, sourceUrl);
         const resolvedPath = this._predictor._sourcePathResolver
-            ? this._predictor._sourcePathResolver.urlToAbsolutePath(resolvedUrl)
-            : urlUtils.fileUrlToAbsolutePath(resolvedUrl);
-        if (resolvedPath)
-          this._addMapping(absolutePath, resolvedPath, url);
+          ? this._predictor._sourcePathResolver.urlToAbsolutePath(resolvedUrl)
+          : urlUtils.fileUrlToAbsolutePath(resolvedUrl);
+        if (!resolvedPath) {
+          continue;
+        }
+
+        let set = sourcePathToCompiled.get(resolvedPath);
+        if (!set) {
+          set = new Set();
+          sourcePathToCompiled.set(resolvedPath, set);
+        }
+        set.add({ ...metadata, sourceUrl: url });
       }
-      map.destroy();
-    } catch (e) {
     }
-  }
 
-  _addMapping(compiledPath: string, sourcePath: string, sourceUrl: string) {
-    let set = this._sourcePathToCompiled.get(sourcePath);
-    if (!set) {
-      set = new Set();
-      this._sourcePathToCompiled.set(sourcePath, set);
-    }
-    set.add({compiledPath, sourceUrl});
+    return sourcePathToCompiled;
   }
 
-  async predictResolvedLocations(params: Dap.SetBreakpointsParams) {
-    await this._done;
+  public async predictResolvedLocations(params: Dap.SetBreakpointsParams) {
+    const sourcePathToCompiled = await this._sourcePathToCompiled;
     const absolutePath = params.source.path!;
-    const set = this._sourcePathToCompiled.get(absolutePath);
-    if (!set)
-      return;
-    for (const {compiledPath, sourceUrl} of set) {
-      const sourceMapUrl = this._sourceMapUrls.get(compiledPath);
-      if (!sourceMapUrl)
+    const set = sourcePathToCompiled.get(absolutePath);
+
+    if (!set) return;
+    for (const metadata of set) {
+      if (!metadata.compiledPath) {
+        return;
+      }
+
+      const map = await sourceUtils.loadSourceMap(metadata);
+      if (!map) {
         continue;
-      try {
-        const map = await sourceUtils.loadSourceMap(sourceMapUrl, 0);
-        if (!map)
+      }
+
+      for (const b of params.breakpoints || []) {
+        const entry = map.generatedPositionFor({
+          source: metadata.sourceUrl,
+          line: b.line,
+          column: b.column || 1,
```

Comment:
Aren't columns in source-maps zero-based? Would this create a problem if we wanted a breakpoint at column 0 (e.g. column would evaluate to false because it's 0, and we'd always swap in 1)?

Reply:
This was just moving around whitespace here, but. DAP columns and lines are 1-based, CDP columns and lines and 0-based. In PWA we keep them 1-based until right before we send the CDP request (in _setByScriptId and _setByUrl in breakpoints.ts)

Reply:
Right, but in the source-map library, lines are 1-based, and columns are 0 based. We have to convert from DAP here.